geode-core/src/test/java/org/apache/geode/internal/tcp/ConnectionTransmissionTest.java

```diff
@@ -0,0 +1,258 @@
+      boolean isSenderInView) {
+    try {
+      final SocketChannel readerSocket = acceptorSocket.accept();
+      final ConnectionTable readerTable = mockConnectionTable();
+      if (isSenderInView) {
+        when(readerTable.getConduit().waitForMembershipCheck(any())).thenReturn(true);
+      }
+
+      final Connection reader = spy(new Connection(readerTable, readerSocket.socket()));
+      CompletableFuture.runAsync(() -> {
+        try {
+          reader.initReceiver();
+        } catch (final RuntimeException e) {
+          e.printStackTrace();
+          throw e;
+        }
+      });
+      return reader;
+    } catch (final Exception e) {
+      e.printStackTrace();
+      throw new RuntimeException(e);
+    }
+  }
+
+  private ConnectionTable mockConnectionTable() throws UnknownHostException {
+    final ConnectionTable connectionTable = mock(ConnectionTable.class);
+    final Distribution distribution = mock(Distribution.class);
+    final DistributionManager distributionManager = mock(DistributionManager.class);
+    final DMStats dmStats = mock(DMStats.class);
+    final CancelCriterion stopper = mock(CancelCriterion.class);
+    final SocketCloser socketCloser = mock(SocketCloser.class);
+    final TCPConduit tcpConduit = mock(TCPConduit.class);
+    final ThreadsMonitoring threadMonitoring = mock(ThreadsMonitoring.class);
+    final AbstractExecutor threadMonitoringExecutor = mock(AbstractExecutor.class);
+    final DistributionConfig config = mock(DistributionConfig.class);
+
+    System.setProperty(SECURITY_SYSTEM_PREFIX + SECURITY_PEER_AUTH_INIT, "true");
+    when(connectionTable.getBufferPool()).thenReturn(new BufferPool(dmStats));
+    when(connectionTable.getConduit()).thenReturn(tcpConduit);
+    when(connectionTable.getDM()).thenReturn(distributionManager);
+    when(distributionManager.getConfig()).thenReturn(config);
+    when(connectionTable.getSocketCloser()).thenReturn(socketCloser);
+    when(distributionManager.getDistribution()).thenReturn(distribution);
+    when(stopper.cancelInProgress()).thenReturn(null);
+    when(tcpConduit.getCancelCriterion()).thenReturn(stopper);
+    when(tcpConduit.getDM()).thenReturn(distributionManager);
+    when(tcpConduit.getSocketId()).thenReturn(new InetSocketAddress(getLocalHost(), 10337));
+    when(tcpConduit.getStats()).thenReturn(dmStats);
+    when(distributionManager.getThreadMonitoring()).thenReturn(threadMonitoring);
+    when(threadMonitoring.createAbstractExecutor(any())).thenReturn(threadMonitoringExecutor);
+    when(tcpConduit.getConfig()).thenReturn(config);
+    tcpConduit.tcpBufferSize = DEFAULT_SOCKET_BUFFER_SIZE;
+
+    doAnswer(invocationOnMock -> {
+      final Runnable runnable = (invocationOnMock.getArgument(0));
+      CompletableFuture.runAsync(() -> {
+        try {
+          runnable.run();
+
+        } catch (final Exception e) {
+          e.printStackTrace();
+          throw e;
+        }
+      });
+      try {
+        Thread.sleep(1000);
```

Comment:
What is the purpose of this `sleep()`? My concern is that it can contribute to nondeterminism in the test.

Could the sleep be replaced with something deterministic like a `CountDownLatch`?

Reply:
I added the sleep to make sure that the handshake reader gets closed (if isCancelInProgress is set to true) before the test continues. Replaced it with sync/async execution of the reader thread.